---
[v1] net: ethernet: mtk_eth_soc: fix panic with napi_enable in mtk_open

Fix kernel panic during late ETH initialization when mtk_open is called before
netif_napi_add completes. This patch reorders NAPI initialization to ensure proper
setup before interface opening, preventing race conditions.
---
v1: Fix panic with napi_enable in mtk_open.
---

From 54298b484f6bd28a33901d2b3562dc536c7807e8 Mon Sep 17 00:00:00 2001
From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Thu, 18 Sep 2025 17:26:48 +0800
Subject: [PATCH] net: ethernet: mtk_eth_soc: fix panic issue with napi_enable 
 in mtk_open

In some scenarios, the ETH may be initialized late, making it very
close to when mtk_open is called. This could lead to a kernel panic at
netif_napi_enable because the ETH driver does not call netif_napi_add().

Signed-off-by: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
---
 drivers/net/ethernet/mediatek/mtk_eth_soc.c | 44 ++++++++++++---------
 1 file changed, 26 insertions(+), 18 deletions(-)

--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -4797,6 +4797,12 @@ static void mtk_pending_work(struct work
 	rtnl_unlock();
 }
 
+static void mtk_napi_del(struct mtk_eth *eth)
+{
+	netif_napi_del(&eth->tx_napi);
+	netif_napi_del(&eth->rx_napi);
+}
+
 static int mtk_free_dev(struct mtk_eth *eth)
 {
 	int i;
@@ -6022,17 +6028,9 @@ static int mtk_probe(struct platform_dev
 
 err_unreg_netdev:
 	mtk_unreg_dev(eth);
+	free_netdev(eth->dummy_dev);
 err_deinit_ppe:
-	/* A netdev opened while probe was still registering can have queued
-	 * the frame-engine reset worker, which takes rtnl and drives the
-	 * hardware; cancel it before anything below is torn down.
-	 */
-	cancel_work_sync(&eth->pending_work);
-	if (eth->dummy_dev) {
-		netif_napi_del(&eth->tx_napi);
-		netif_napi_del(&eth->rx_napi);
-		free_netdev(eth->dummy_dev);
-	}
+	mtk_napi_del(eth);
 	mtk_ppe_deinit(eth);
 	mtk_release_all_muxes(eth);
 	mtk_mdio_cleanup(eth);
@@ -6071,8 +6069,7 @@ static void mtk_remove(struct platform_d
 	mtk_wed_exit();
 	mtk_hw_deinit(eth);
 
-	netif_napi_del(&eth->tx_napi);
-	netif_napi_del(&eth->rx_napi);
+	mtk_napi_del(eth);
 	mtk_cleanup(eth);
 	free_netdev(eth->dummy_dev);
 	mtk_mdio_cleanup(eth);
